identity/src/main/java/com/android/identity/GattServer.java

```diff
@@ -209,25 +268,103 @@ public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, i
             } catch (SecurityException e) {
                 reportError(e);
             }
+        } else if (mCharacteristicL2CAP != null
+                && characteristic.getUuid().equals(mCharacteristicL2CAPUuid)) {
+            if (!mSupportL2CAP) {
+                reportError(new Error("Unexpected read request for L2CAP characteristic, not supported"));
+                return;
+            }
+            if (mServerSocket == null) {
+                reportError(new Error("Socket for L2CAP not available"));
+                return;
+            }
+            int psm = mServerSocket.getPsm();
+            byte[] psmValue = ByteBuffer.allocate(4).putInt(psm).array();
+            if (mLog.isTransportEnabled()) {
+                mLog.transport("PSM Value generated: " + psm + " byte: " + Util.toHex(psmValue));
+            }
+            mGattServer.sendResponse(device,
+                    requestId,
+                    BluetoothGatt.GATT_SUCCESS,
+                    0,
+                    psmValue);
+            acceptConnection();
         } else {
             reportError(new Error("Read on unexpected characteristic with UUID "
                     + characteristic.getUuid()));
         }
     }
 
+    private void acceptConnection() {
+        try {
+            mSocket = mServerSocket.accept();
```

Comment:
So accept() will block until a connection is made... the only reason this works is that you are calling it right after the L2CAP characteristic is read and hoping the GATT client will try to connect after this...

I think the correct way to deal with this is to create a thread for listening for incoming connections right after you create `mServerSocket`... and when you get a connection you will reuse that thread to do I/O.

Reply:
Thread created for L2CAP socket